store/store.go

```diff
@@ -546,6 +559,110 @@ type requestOptions struct {
 	Data         []byte
 }
 
+type httpErrorResponse struct {
+	StatusCode int
+	URL        *url.URL
+	Method     string
+	Oops       string
+}
+
+func NewHttpErrorResponse(resp *http.Response) *httpErrorResponse {
+	return &httpErrorResponse{StatusCode: resp.StatusCode, URL: resp.Request.URL, Method: resp.Request.Method}
+}
+
+func (r httpErrorResponse) Error() string {
+	return fmt.Sprintf("HTTP request failed: '%v', method: %s, status: %d", r.URL, r.Method, r.StatusCode)
+}
+
+func (s *Store) doStoreRequest(client *http.Client, req *http.Request, user *auth.UserState, dataDecodeFunc func(*http.Response) error) error {
+	var err error
+	for refreshed := true; refreshed; {
+		refreshed = false
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+
+		wwwAuth := resp.Header.Get("WWW-Authenticate")
+		if resp.StatusCode == http.StatusUnauthorized {
+			if user != nil && strings.Contains(wwwAuth, "needs_refresh=1") {
+				// refresh user
+				err := s.refreshUser(user)
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+				refreshed = true
+			}
+			if strings.Contains(wwwAuth, "refresh_device_session=1") {
+				// refresh device session
+				if s.authContext == nil {
+					resp.Body.Close()
+					return fmt.Errorf("internal error: no authContext")
+				}
+				device, err := s.authContext.Device()
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+
+				err = s.refreshDeviceSession(device)
+				if err != nil {
+					resp.Body.Close()
+					return err
+				}
+				refreshed = true
+			}
+		}
+
+		if !refreshed {
+			err = dataDecodeFunc(resp)
+			if err != nil {
+				return err
+			}
+		}
+		resp.Body.Close()
+	}
+	return err
+}
+
+func isFatal(err error) bool {
+	if httpStatusErr, ok := err.(httpErrorResponse); ok {
+		if httpStatusErr.StatusCode == http.StatusInternalServerError || httpStatusErr.StatusCode == http.StatusServiceUnavailable {
+			return false
+		}
+	} else if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+		return false
+	} else if err == io.ErrUnexpectedEOF {
+		return false
+	}
+	return true
+}
+
+// retryRequest does an authenticated request to the store handling a potential macaroon refresh required if needed.
+// It retries the request on certain http errors or unexpected EOFs.
+func (s *Store) retryRequest(client *http.Client, reqOptions *requestOptions, user *auth.UserState, dataDecodeFunc func(*http.Response) error) error {
+	req, err := s.newRequest(reqOptions, user)
+	if err != nil {
+		return err
+	}
+
+	err = retry.Call(retry.CallArgs{
+		Func: func() error {
+			return s.doStoreRequest(client, req, user, dataDecodeFunc)
+		},
+		Attempts:     6,
+		Delay:        time.Millisecond,
+		BackoffFunc:  retry.DoubleDelay,
+		MaxDuration:  MaxRetryDuration,
+		Clock:        clock.WallClock,
```

Comment:
As discussed on the standup the clock will most probably reflect the system clock after resume, meaning the retry mechanism will stop retrying as MaxRetryDuration will be immediately exceeded. I don't think this is a problem we should be worried about given that after resume the networking needs to be brought up and this will exceed any of our retry delays for sure.